src/gui/src/tclCmdInputWidget.cpp

```diff
@@ -0,0 +1,239 @@
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "tclCmdInputWidget.h"
+
+#include <tcl.h>
+
+#include <QMimeData>
+#include <QTextStream>
+
+namespace gui {
+
+TclCmdInputWidget::TclCmdInputWidget(QWidget* parent) :
+    QPlainTextEdit(parent)
+{
+  setObjectName("tcl_scripting");  // for settings
+  setPlaceholderText("TCL commands");
+  setAcceptDrops(true);
+
+  max_height_  = QWIDGETSIZE_MAX;
+  determineLineHeight();
+
+  highlighter_ = nullptr;
+
+  // precompute size for updating text box size
+  document_margins_ = 2 * document()->documentMargin();
+
+  connect(this, SIGNAL(textChanged()), this, SLOT(updateSize()));
+  updateSize();
+}
+
+TclCmdInputWidget::~TclCmdInputWidget()
+{
+}
+
+void TclCmdInputWidget::setFont(const QFont& font)
+{
+  QPlainTextEdit::setFont(font);
+
+  determineLineHeight();
+}
+
+void TclCmdInputWidget::determineLineHeight()
+{
+  QFontMetrics font_metrics = fontMetrics();
+  line_height_ = font_metrics.lineSpacing();
+
+  double tab_indent_width = 2 * font_metrics.averageCharWidth();
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 10, 0))
+  // setTabStopWidth deprecated in 5.10
+  setTabStopDistance(tab_indent_width);
+#else
+  setTabStopWidth(tab_indent_width);
+#endif
+}
+
+void TclCmdInputWidget::keyPressEvent(QKeyEvent* e)
+{
+  bool forward_key_press = true;
+
+  bool has_control = e->modifiers().testFlag(Qt::ControlModifier);
+  int key = e->key();
+  if (key == Qt::Key_Enter || key == Qt::Key_Return) {
+    // Handle enter
+    if (has_control) {
+      // don't execute just insert the newline
+      // does not get inserted by Qt, so manually inserting
+      insertPlainText("\n");
```

Comment:
From our side discussion:
maliberty 2 hours ago 
the command seems to execute when ctrl-enter is hit even though it seems the purpose is to allow a multi-line command to be entered.

 
@arlpetergadfort arlpetergadfort 10 minutes ago Author
You are correct that it should insert a newline at the cursor and not attempt to check the command. I wonder what is different with your setup from mine?. If I don't insert the newline, Qt seems to ignore the control+enter command.

 
@maliberty maliberty 5 minutes ago 
So you don't see it executing the command? Perhaps it relates to the key release. I'll try playing around with it later and see if I can find the source.

 
@arlpetergadfort arlpetergadfort 2 minutes ago Author
nope, for me it correctly inserts the command I found that qInfo() << e; is helpful to see the actual key event, although did see some messages online about the modifier keys not always showing up correctly. I rebased it so you may need to re-checkout the branch again.

Reply:
This seems ok now - I'm not sure if you fixed it or if it was pilot error before.